pennylane_qiskit/devices.py

```diff
@@ -171,20 +191,25 @@ def apply(self, operation, wires, par):
         else:
```

Comment:
This part of the code is a bit confusing. Particularly, this error should be raised higher up I think. We want to see if the user has supplied a correct operation which is mapped. So something like

```
device.apply("RXY", wires=[1])
```
should raise an error since it is not a mapped operation. But with the code now, it will raise a key error as 

```
Traceback (most recent call last):
  File "/Users/shahnawaz/Dropbox/dev/pennylane-qiskit/tests/test_unsupported_operations.py", line 67, in test_unsupported_operation
    device.apply('RXY', wires=[0], par=[0.1])
  File "/Users/shahnawaz/Dropbox/dev/pennylane-qiskit/pennylane_qiskit/devices.py", line 170, in apply
    mapped_operation = self._operation_map[operation]
KeyError: 'RXY'
```

So, I suggest the check to happen before we even do `mapped_operation = self._operation_map[operation]` as,
```
try:
    mapped_operation = self._operation_map[operation]
except KeyError:
    raise ValueError("")
```

Reply:
Also, if we make sure that the operation is one of the supported ones in `QISKIT_OPERATION_MAP`, we do not need to check beyond the two possible types that we allow in the map. Therefore, I am adding an error message and checking if the operation really is valid in the beginning with something like:

```
ValueError: The operation is not of an expected type. Supported QISKIT operations and instructions are: PauliX, PauliY, PauliZ, CNOT, CZ, SWAP, RX, RY, RZ, PhaseShift, Hadamard, Rot, BasisState, QubitStateVector, QubitUnitary, S, T, U1, U2, U3
```

Reply:
Something to note: the `Device` base class in PennyLane currently has built-in checks for supported observables and operations. When `Device.execute()` is called, before any `apply()` calls are made, the device calls `Device.check_validity` (see here: https://github.com/XanaduAI/pennylane/blob/478e0fa6fd692768105002d8a09d99e70ae7644b/pennylane/_device.py#L303).

This checks if the queued operations and observables are in the `Device.operations` and `Device.observables` properties; if not, an exception is raised.

Reply:
Thanks. Then, probably we should make the check happen before the `.apply` method is called in the `check_validity` function.